// waves-api/raw/src/classes/Money.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var bignumber_js_1 = require("bignumber.js");
var Asset_1 = require("./Asset");
function checkAmount(amount) {
    if (!(typeof amount === 'string' || amount instanceof bignumber_js_1.default)) {
        throw new Error('Please use strings to create instances of Money');
    }
}
function getDivider(precision) {
    return new bignumber_js_1.default(10).pow(precision);
}
function getAsset(asset) {
    if (typeof asset === 'string') {
        return Asset_1.default.get(asset);
    }
    else {
        return Promise.resolve(Asset_1.default.create(asset));
    }
}
var Money = /** @class */ (function () {
    function Money(coins, asset) {
        if (!Asset_1.default.isAsset(asset)) {
            throw new Error('Please use Asset for the `asset` argument');
        }
        this.asset = asset;
        this.coins = new bignumber_js_1.default(coins);
        this.divider = getDivider(asset.precision);
    }
    Money.prototype.toCoins = function () {
        return this.coins.toFixed(0);
    };
    Money.prototype.toTokens = function () {
        return this.coins.div(this.divider).toFixed(this.asset.precision);
    };
    Money.prototype.toJSON = function () {
        return {
            assetId: this.asset.id,
            tokens: this.toTokens()
        };
    };
    Money.prototype.toString = function () {
        return this.toTokens() + " " + this.asset.id;
    };
    return Money;
}());
exports.default = {
    fromCoins: function (coins, asset) {
        checkAmount(coins);
        return getAsset(asset).then(function (a) {
            return new Money(coins, a);
        });
    },
    fromTokens: function (tokens, asset) {
        checkAmount(tokens);
        return getAsset(asset).then(function (a) {
            var divider = getDivider(a.precision);
            var coins = new bignumber_js_1.default(tokens).mul(divider).toFixed(0);
            return new Money(coins, a);
        });
    },
    isMoney: function (object) {
        return object instanceof Money;
    }
};
